### ISSUE TITLE
[feature] Team 2 - Minion enemy movement 

### ISSUE THREAD

**Issue by FleurN21** [UQdeco2800/2021-studio-2#5]
# Description 
This will include basic enemy design, movement, and collision detection. Multiple enemy textures and shaders are needed to give the game more diverse enemies without needing to be redesigned completely. For initial purposes, the enemies will start from a set location until the map design team are ready to set up possible (randomised) enemy spawning locations. The enemy will wander around the room until they meet the player so they can attack them without bee-lining directly for the player. The enemy will approach the enemy until they are ready to attack. The enemy will prefer to stay orbiting the player at its attack range so ranged enemies can attack the player without getting too close. Enemies will bounce off from objects on the map, the walls, and each other. Eventually, this will be modified to make the enemy turn and avoid walking into walls but is only just groundwork for now. Enemies may bounce off of the player depending on how melee attacking is implemented by Team 5.

Once this is done we will attempt to add more features. Two features being considered but may be cut are enemies spreading out around the player before attacking which will also tie into spreading around the map to cover more ground. If this was added it would also be necessary to add a call out ability. If an enemy isn't killed fast enough, when it detects the player, it will call out to nearby enemies for assistance. A hive-based enemy is also being planned on, as such it would be good to allow certain enemies to hover around a set location and not to be pulled away from that location by the player. As such it will remain within the set boundary and will stop pursuing the player if pulled out of this area. When ranged enemies are added it would also be preferred to keep them at a range. Ranged enemies would instead approach the player until they can attack, at which point it will keep its distance at its attack range to be safe from the player's melee attacks.

Some of these features will rely on other teams. It will be apart of the map design as to where enemies will spawn and what enemies spawn in what areas. We also need to talk to the attacks team to find the ranges of melee and ranged attacks and if there is any knockback from certain attacks. Finally, we need to make sure the enemy and player movement styles match to prevent issues such as the player moving smoothly in all areas and the enemy moving "jumps" a time and only in set directions.

#  Milestones 
## Design: 
- [x] #15
- [x] #16

## Placement:
- [x] #17

## Movement
- [x] #18
- [x] #19
- [x] #43

## Attacks
- [x] #45

## Collision
- [x] #20
- [x] #21
- [x] #22
- [x] #44

## Stretch Goals: 
- Enemies avoid each other to take up more space, then surround the player on attack (possibly)
- [x] #60
- Create more enemy art and behaviour ideas for the next sprint (ranged and hive based)
- [x] #57
- [x] #58
- Various arrow effects (multishot, tracking arrow, sniper arrow, projectile path warning)

# Dependencies
- Collaborate with player movement (Team 1) to ensure the player and NPC movement's don't conflict #4 
- Collaborate/integrate with map team (Team 4) for collisions and positioning of walls, objects, and enemy spawn placement #6 
    - Also need a testing arena with different scenarios to test enemy movement
- Collaborate/integrate with hero health, attacks, and position (Team 5)  for interacting with the player #3 

# Members
- Fleur Nicholson (@FleurN21) Fleur#0608
- Minh Tieu (@MTbir) Minh#7523
- Eli Cox (@eco9898) .♡.𝓮𝓵𝓲.♡.#5723
- Kai Barry (@BobJohnsom) BOBJohnson#5476
- Haopeng Xia (@HaoPengXia) HaoPengXia(Barney)#8321

**Comment by eco9898** [comment]
Accidentally closed

**Comment by eco9898** [comment]
Added a description to the feature ticket justifying each feature as required.

## Milestone changes:
Added a design milestone:
 - Create different textures/shaders the enemy can spawn with (just cosmetic)
Moved a movement milestone "Enemies approach the player" from stretch goals to main features as it is already implemented but needs modification
Modified a movement milestone "Enemies approach the player" to stop at attack range
 - Enemies approach the player, when they are near, until within attack range (either ranged or melee depending on the enemy type)
Moved a movement milestone "collision with the player" from stretch goals to main features as it is already implemented but may need to be adjusted to the requirements of our combat system

## Stretch Goals changes:
Added new stretch goals:
 - Enemies avoid each other to take up more space, then surround the player on attack (possibly)
 - Enemies can call out to nearby enemies when they find the player (possibly)
 - Create more enemy art and behavior ideas for the next sprint (ranged and hive based)
 - Some enemies "hover" in a set area and prefer to stay there then chase the player (they will return if too far away)
 - Enemies will orbit at their attack range so ranged enemies can attack from afar, even as the player approaches them.

## Dependency changes:
Added ticket links to the dependencies of team 1 and 5
Modified "hero health, attacks, and position" dependency from team 3 to team 5 as team 3 is only implementing stylistic low health features and team 5 is implementing combat and attacking.
Added "player movement" dependency with team 1 to make sure we both implement character movement similarly as they will work on player movement.
Added "Testing arena" dependency with team 4 as we will need a mock-up map to test how our enemies interact with the map correctly.

**Comment by zenzlia** [comment]
Please make sure to create any task tickets for this feature

**Comment by eco9898** [comment]
Added new stretch goal:
- Make a projectile system which makes the targets the player from an enemy

**Comment by eco9898** [comment]
Added #44 and #43

**Comment by eco9898** [comment]
Completed extra arrow feature multishot in branch team-2
